### PR TITLE
Add spdk_bdev_passthru() to pass through nvme admin commands.

### PR DESCRIPTION
**Pull request by bsedyang**

### CODE REVIEW

File: include/spdk/bdev.h [hunk 1]
```diff
@@ -133,6 +133,7 @@ enum spdk_bdev_io_type {
 	SPDK_BDEV_IO_TYPE_UNMAP,
 	SPDK_BDEV_IO_TYPE_FLUSH,
 	SPDK_BDEV_IO_TYPE_RESET,
+	SPDK_BDEV_IO_TYPE_PASSTHRU,
```
Comment (dverkamp-intel) [inline include/spdk/bdev.h]: This should be SPDK_BDEV_IO_TYPE_NVME_PASSTHRU in case we want to pass through other command sets later.

It also might be a good idea to make I/O and admin commands two separate I/O types so that bdev modules can support just one or the other.

File: include/spdk/bdev.h [hunk 2]
```diff
@@ -229,6 +230,8 @@ struct spdk_bdev_io *spdk_bdev_unmap(struct spdk_bdev *bdev, struct spdk_io_chan
 struct spdk_bdev_io *spdk_bdev_flush(struct spdk_bdev *bdev, struct spdk_io_channel *ch,
 				     uint64_t offset, uint64_t length,
 				     spdk_bdev_io_completion_cb cb, void *cb_arg);
+struct spdk_bdev_io *spdk_bdev_passthru(struct spdk_bdev *bdev, void *cmd, void *buf,
```
Comment (dverkamp-intel) [inline include/spdk/bdev.h]: This should be named spdk_bdev_nvme_passthru() to match the I/O type, as mentioned above (and if we decide to make admin and I/O two separate I/O types, also include that as part of the passthru function name).

cmd should be declared as `const struct spdk_nvme_cmd *`.

File: include/spdk/bdev.h [hunk 3]
```diff
@@ -229,6 +230,8 @@ struct spdk_bdev_io *spdk_bdev_unmap(struct spdk_bdev *bdev, struct spdk_io_chan
...
+					uint16_t nbytes, spdk_bdev_io_completion_cb cb, void *cb_arg);
```
Comment (dverkamp-intel) [inline include/spdk/bdev.h]: Any particular reason nbytes is a uint16_t here?  I think a size_t would be better as a buffer size.

File: include/spdk_internal/bdev.h [hunk 4]
```diff
@@ -213,6 +213,12 @@ struct spdk_bdev_io {
 		struct {
 			enum spdk_bdev_reset_type type;
 		} reset;
+		struct {
+			void *cmd;
```
Comment (dverkamp-intel) [inline include/spdk_internal/bdev.h]: We should probably just make a copy of the command here rather than using a pointer (so that the user doesn't have to hold on to a copy).

If we do keep this as a pointer, the lifetime requirements of cmd definitely need to be documented in the Doxygen comment for the passthru function.

File: lib/bdev/bdev.c [hunk 5]
```diff
@@ -412,6 +412,8 @@ __submit_request(struct spdk_bdev *bdev, struct spdk_bdev_io *bdev_io)
 	if (bdev_io->type == SPDK_BDEV_IO_TYPE_RESET) {
 		spdk_bdev_cleanup_pending_buf_io(bdev);
 		ch = NULL;
+	} else if (bdev_io->type == SPDK_BDEV_IO_TYPE_PASSTHRU) {
+		ch = NULL;
```
Comment (dverkamp-intel) [inline lib/bdev/bdev.c]: Any reason this can't use the normal I/O channel mechanism?

File: lib/bdev/nvme/blockdev_nvme.c [hunk 6]
```diff
@@ -969,6 +978,36 @@ bdev_nvme_queued_done(void *ref, const struct spdk_nvme_cpl *cpl)
...
+	if (sct == SPDK_NVME_SCT_GENERIC && sc == SPDK_NVME_SC_SUCCESS) {
+		bdev_io->status = SPDK_BDEV_IO_STATUS_SUCCESS;
+	} else {
+		bdev_io->error.nvme.sct = sct;
+		bdev_io->error.nvme.sc = sc;
+		bdev_io->status = SPDK_BDEV_IO_STATUS_NVME_ERROR;
+	}
+
+	spdk_event_call(spdk_event_allocate(bdev_io->u.passthru.p_lcore,
+					    (spdk_event_fn)bdev_nvme_passthru_completion, bdev_io,
```
Comment (dverkamp-intel) [inline lib/bdev/nvme/blockdev_nvme.c]: This should not need a function pointer cast (remove `(spdk_event_fn)`).

File: lib/bdev/nvme/blockdev_nvme.c [hunk 7]
```diff
@@ -969,6 +978,36 @@ bdev_nvme_queued_done(void *ref, const struct spdk_nvme_cpl *cpl)
...
+					    (void *)bdev_io->status));
```
Comment (dverkamp-intel) [inline lib/bdev/nvme/blockdev_nvme.c]: It might be easier to just preserve the entire spdk_nvme_cpl in the bdev_io union and then call spdk_bdev_io_complete_nvme_status() in the event function.

File: lib/nvme/nvme_ns.c [hunk 8]
```diff
@@ -163,6 +163,7 @@ spdk_nvme_ns_get_flags(struct spdk_nvme_ns *ns)
 
 enum spdk_nvme_pi_type
 spdk_nvme_ns_get_pi_type(struct spdk_nvme_ns *ns) {
+
```
Comment (dverkamp-intel) [inline lib/nvme/nvme_ns.c]: Unrelated whitespace change?

File: include/spdk/bdev.h (hunk 1 above)
Comment (benlwalker) [inline include/spdk/bdev.h]: Just define two values in this enum instead of SPDK_BDEV_IO_TYPE_PASSTHRU:

SPDK_BDEV_IO_TYPE_NVME_ADMIN
SPDK_BDEV_IO_TYPE_NVME_IO

File: include/spdk_internal/bdev.h (hunk 4 above)
Comment (benlwalker) [inline include/spdk_internal/bdev.h]: I agree with Daniel - don't store a pointer and store a struct spdk_nvme_cmd directly. Make a copy from the user's command.

File: include/spdk/bdev.h (hunk 1 above)
Comment (pvonstamwitz) [inline include/spdk/bdev.h]: Let's limit this PR to SPDK_BDEV_IO_TYPE_NVME_ADMIN only for now. I am not convinced of the need to provide a pass-through for IO, but we could add it later is needed.

File: lib/bdev/bdev.c (hunk 5 above)
Comment (pvonstamwitz) [inline lib/bdev/bdev.c]: For SPDK_BDEV_IO_TYPE_NVME_IO, we would use the channel, but it is not required for SPDK_BDEV_IO_TYPE_NVME_ADMIN.

File: include/spdk/bdev.h (hunk 1 above)
Comment (benlwalker) [inline include/spdk/bdev.h]: Good with me

Comment (bsedyang) [inline include/spdk/bdev.h]: Thanks. Will do it.

File: include/spdk/bdev.h (hunk 3 above)
Comment (bsedyang) [inline include/spdk/bdev.h]: will change to size_t

File: include/spdk_internal/bdev.h (hunk 4 above)
Comment (bsedyang) [inline include/spdk_internal/bdev.h]: Thanks for advising.

File: lib/bdev/bdev.c (hunk 5 above)
Comment (bsedyang) [inline lib/bdev/bdev.c]: agree with Paul..

File: lib/bdev/nvme/blockdev_nvme.c (hunk 7 above)
Comment (bsedyang) [inline lib/bdev/nvme/blockdev_nvme.c]: Daniel, thank you very much for your advice. Will change it,

File: include/spdk_internal/bdev.h [hunk 9]
```diff
@@ -271,6 +271,13 @@ struct spdk_bdev_io {
 		struct {
 			enum spdk_bdev_reset_type type;
 		} reset;
+		struct {
+			struct spdk_nvme_cmd cmd;
+			void *buf;
+			size_t nbytes;
+			uint32_t p_lcore;
+			struct spdk_nvme_cpl cpl;
```
Comment (benlwalker) [inline include/spdk_internal/bdev.h]: Copying the completion is an implementation detail of the blockdev_nvme module, so put a copy of the spdk_nvme_cpl in struct nvme_bdev_io instead of here.

File: include/spdk_internal/bdev.h [hunk 10]
```diff
@@ -271,6 +271,13 @@ struct spdk_bdev_io {
 		struct {
 			enum spdk_bdev_reset_type type;
 		} reset;
+		struct {
+			struct spdk_nvme_cmd cmd;
+			void *buf;
+			size_t nbytes;
+			uint32_t p_lcore;
```
Comment (benlwalker) [inline include/spdk_internal/bdev.h]: The lcore is used to pass a message back to the correct core, but the existence of the message pass is purely an implementation detail of the blockdev_nvme module. Instead, store the lcore in struct nvme_bdev_io.

Or, you can probably store a struct spdk_event * in the nvme_bdev_io that you allocate at submission time. That way if the event allocation fails, you can immediately fail the command back to the user.

File: lib/bdev/bdev.c [hunk 11]
```diff
@@ -857,6 +859,38 @@ spdk_bdev_reset(struct spdk_bdev *bdev, enum spdk_bdev_reset_type reset_type,
...
+	uint32_t p_lcore = spdk_env_get_current_core();
+
+	bdev_io = spdk_bdev_get_io();
+	if (!bdev_io) {
+		SPDK_ERRLOG("bdev_io memory allocation failed during nvme_admin_passthru\n");
+		return NULL;
+	}
+
+	bdev_io->type = SPDK_BDEV_IO_TYPE_NVME_ADMIN;
+	memcpy(&bdev_io->u.nvme_admin_passthru.cmd, cmd, sizeof(struct spdk_nvme_cmd));
```
Comment (benlwalker) [inline lib/bdev/bdev.c]: This is better written as a direct assignment because then you can't get the size wrong.

bdev_io->u.nvme_admin_passthru.cmd = *cmd;

File: lib/bdev/nvme/blockdev_nvme.c [hunk 12]
```diff
@@ -969,6 +979,27 @@ bdev_nvme_queued_done(void *ref, const struct spdk_nvme_cpl *cpl)
...
+					  bdev_io->u.nvme_admin_passthru.cpl.status.sct,
+					  bdev_io->u.nvme_admin_passthru.cpl.status.sc);
+}
+
+static void
+bdev_nvme_admin_passthru_done(void *ref, const struct spdk_nvme_cpl *cpl)
+{
+	struct spdk_bdev_io *bdev_io = spdk_bdev_io_from_ctx((struct nvme_bdev_io *)ref);
+
+	memcpy(&bdev_io->u.nvme_admin_passthru.cpl, cpl, sizeof(*cpl));
```
Comment (benlwalker) [inline lib/bdev/nvme/blockdev_nvme.c]: This is better written as an assignment because then you can't get the size wrong.

bdev_io->u.nvme_admin_passthru.cpl = *cpl;

File: lib/bdev/nvme/blockdev_nvme.c [hunk 13]
```diff
@@ -969,6 +979,27 @@ bdev_nvme_queued_done(void *ref, const struct spdk_nvme_cpl *cpl)
...
+
+	spdk_event_call(spdk_event_allocate(bdev_io->u.nvme_admin_passthru.p_lcore,
```
Comment (benlwalker) [inline lib/bdev/nvme/blockdev_nvme.c]: While unlikely, spdk_event_allocate can actually fail in out of memory scenarios. I think you probably want to allocate the event at submission time and store a pointer to it in struct nvme_bdev_io. That way if the allocation fails you can immediately return to the user without executing the command.

File: lib/bdev/nvme/blockdev_nvme.c [hunk 14]
```diff
@@ -92,6 +92,15 @@ struct nvme_bdev_io {
 
 	/** Offset in current iovec. */
 	uint32_t iov_offset;
+
+	/** Saved status for admin passthru completion event. */
+	struct spdk_nvme_cpl cpl;
+
```
Comment (benlwalker) [inline lib/bdev/nvme/blockdev_nvme.c]: The lcore is stored inside of the event, so you can delete this p_lcore member.

File: lib/bdev/bdev.c [hunk 15]
```diff
@@ -857,6 +859,36 @@ spdk_bdev_reset(struct spdk_bdev *bdev, enum spdk_bdev_reset_type reset_type,
 	return rc;
 }
 
+struct spdk_bdev_io *
```
Comment (benlwalker) [inline lib/bdev/bdev.c]: I know NVMe admin passthru does not need an I/O channel, but for consistency with the other commands I'd prefer we require one in the interface. That unifies all of the code paths.

We just recently added an I/O channel to the reset interface for this same reason.

Comment (pvonstamwitz) [inline lib/bdev/bdev.c]: We'll add the channel to the interface, but reset still resets channel to NULL in __submit_request. Should we keep the same logic for passthru or take it out?

Comment (pvonstamwitz) [inline lib/bdev/bdev.c]: It works whether we call submit_request with ch = NULL or = bdev_io->ch->channel since we don't use it. But we have used the reset code as a model for a command that does not need an I/O channel, so we prefer to set it to NULL to be consistent with that model. Any thoughts?

**Comment by bsedyang** [comment]
Hi Ben, I just pushed new commit but the check failed.
From Details, I couldn't figure out what's wrong.
Would you help me on the issue?  Thanks.

**Comment by dverkamp-intel** [comment]
The Travis check failure was due to a bug in the blobfs unit test (unrelated to your change); this is fixed in commit 1285f823e4f5995bbc423df190d4a326f425873a, so if you rebase your change, it should work now.

**Comment by bsedyang** [comment]
Daniel,  Just did rebase and all checks passed.  Thanks.
